src/main/java/net/openhft/chronicle/queue/reader/ChronicleReader.java

```diff
@@ -123,45 +125,23 @@ public void execute() {
                     try {
                         moveToSpecifiedPosition(queue, tailer, isFirstIteration);
                         lastObservedTailIndex = tailer.index();
-                        Consumer<String> messageConsumer = text -> applyFiltersAndLog(text, tailer.index(), matchCounter);
-                        BooleanSupplier readOne;
-                        if (methodReaderInterface == null) {
-                            readOne = () -> readOne(messageConverter, tailer, messageConsumer);
-                        } else {
-                            // TODO: consider unifying this with messageConverter
-                            Bytes<ByteBuffer> bytes = Bytes.elasticHeapByteBuffer(256);
-                            Wire wire = wireType.apply(bytes);
-                            if (wire instanceof TextWire)
-                                ((TextWire)wire).useTextDocuments();
-                            MethodWriterBuilder<?> mwb = wire.methodWriterBuilder(methodReaderInterface);
-                            if (showMessageHistory)
-                                mwb.updateInterceptor((methodName, t) -> {
-                                    MessageHistory messageHistory = MessageHistory.get();
-                                    // this is an attempt to recognise that no MH was read and instead the method reader called reset(...) on it
-                                    if (messageHistory.sources() != 1 || messageHistory.timings() != 1)
-                                        bytes.append(messageHistory + System.lineSeparator());
-                                    return true;
-                                });
-                            MethodReader methodReader = tailer.methodReader(mwb.build());
-                            readOne = () -> {
-                                boolean found = methodReader.readOne();
-                                if (found)
-                                    messageConsumer.accept(bytes.toString());
-                                bytes.clear();
-                                return found;
-                            };
-                        }
 
                         while (!Thread.currentThread().isInterrupted()) {
-                            boolean found = readOne.getAsBoolean();
+                            if (shouldHaltReadingDueToContentBasedLimit(tailer)) {
+                                running = false;
+                                break;
+                            }
```

Comment:
This isn't ideal because we read every message in the queue twice; once to determine whether we've reached the content based limit, then again to print the entry to the sink. 

The reason for this is the MethodReader never gives us access to the Wire, I could refactor it so we use Wire.methodReader(,,) with the readSkip(-4) trick instead of using a tailer.methodReader() and we might be able to combine the content-based limit and the conversion to a string against the same wire, resetting the positions inbetween.

Reply:
I wonder if there's a less hacky way to do readDocument() on a DocumentContext wire than calling readSkip(-4), it kind of feels brittle or hacky.

Reply:
The more I think about it, the less I care about reading each entry twice. Performance here isn't critical, it's really not much different to reading the enclosed Wire twice anyway, it means we don't have to do any hacky stuff, and it gives us more flexibility in how we implement QueueEntryReaders